### ISSUE TITLE
GraalVM Support

### ISSUE THREAD

**Issue by ahelmbr** [jline/jline3#381]
Is it planned to support usage of jline3 within  GraalVMs native-image?

**Comment by gnodet** [comment]
I'd like to, the problem is that the demo is relying a lot on reflection and makes things a bit difficult to try in GraalVM.  I'd need a simpler demo to experiment.

**Comment by KangoV** [comment]
Just an example of the error I'm getting with GraalVM native image:
```
WARNING: Unable to retrieve infocmp for type dumb
java.lang.NullPointerException
        at java.io.Reader.<init>(Reader.java:78)
        at org.jline.utils.InputStreamReader.<init>(InputStreamReader.java:135)
        at org.jline.utils.InfoCmp.loadDefaultInfoCmp(InfoCmp.java:608)
        at org.jline.utils.InfoCmp.lambda$static$3(InfoCmp.java:619)
        at org.jline.utils.InfoCmp.getLoadedInfoCmp(InfoCmp.java:557)
        at org.jline.utils.InfoCmp.getInfoCmp(InfoCmp.java:545)
        at org.jline.terminal.impl.AbstractTerminal.parseInfoCmp(AbstractTerminal.java:187)
        at org.jline.terminal.impl.DumbTerminal.<init>(DumbTerminal.java:88)
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:401)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:259)
```

**Comment by tim-patterson** [comment]
Hey so the fix for these errors is simply creating a `resource-config.json` file.
See here:
https://github.com/tim-patterson/jsonsql/blob/master/src/main/resources/META-INF/native-image/resource-config.json

**Comment by mattirn** [comment]
I have created a JLine REPL app with builtin commands and tried to produce GraalVM native-image https://github.com/mattirn/jline3/commit/3906925164d08416daad564150b15937db4e46e1. At finally I have succeeded to create it but I had to disable `java.util.logging`. In addition to the logging I encountered a couple of more errors:

**STEP I**
```
                <configuration>
                    <skip>false</skip>
                    <imageName>graal</imageName>
                    <mainClass>org.jline.demo.Graal</mainClass>
                    <buildArgs>
                        --no-fallback
                        -H:ReflectionConfigurationFiles=reflection-config.json
                    </buildArgs>
                </configuration>
```
=>
[graal:18919]     analysis:  12,344.72 ms
Fatal error: com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing com.oracle.svm.reflect.ClassLoader_findLibrary_147c61d6d7f79efa3da939184845f679c7833707.invoke(java.lang.Object, java.lang.Object[]) 
Parsing context:
	parsing java.lang.reflect.Method.invoke(Method.java:498)
	parsing **org.jline.utils.Signals.doRegister(Signals.java:100)**
	parsing org.jline.utils.Signals.registerDefault(Signals.java:67)
	parsing org.jline.terminal.impl.PosixSysTerminal.<init>(PosixSysTerminal.java:48)
	parsing org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:364)
	parsing org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:263)
	parsing org.jline.demo.Graal.main(Graal.java:234)


**STEP II**
```
                <configuration>
                    <skip>false</skip>
                    <imageName>graal</imageName>
                    <mainClass>org.jline.demo.Graal</mainClass>
                    <buildArgs>
                        --no-fallback
                        --report-unsupported-elements-at-runtime
                        -H:ReflectionConfigurationFiles=reflection-config.json
                    </buildArgs>
                </configuration>
```
=>
[graal:18919]     analysis:  17,345.40 ms
Error: Unsupported features in 5 methods
Detailed message:
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: **org.apache.sshd.common.util.security.bouncycastle.BouncyCastleRandom.<init>()**. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Trace: 
	at parsing org.apache.sshd.common.util.security.bouncycastle.BouncyCastleRandomFactory.create(BouncyCastleRandomFactory.java:43)
Call path from entry point to org.apache.sshd.common.util.security.bouncycastle.BouncyCastleRandomFactory.create(): 
	at org.apache.sshd.common.util.security.bouncycastle.BouncyCastleRandomFactory.create(BouncyCastleRandomFactory.java:43)
	at org.apache.sshd.common.util.security.bouncycastle.BouncyCastleRandomFactory.create(BouncyCastleRandomFactory.java:28)
	at org.apache.sshd.common.Factory.get(Factory.java:34)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:188)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:141)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:342)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: **org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderUtils.getEDDSAPrivateKeyType()**. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Trace: 
	at parsing org.apache.sshd.common.util.security.SecurityUtils.getEDDSAPrivateKeyType(SecurityUtils.java:571)
Call path from entry point to org.apache.sshd.common.util.security.SecurityUtils.getEDDSAPrivateKeyType(): 
	no path found from entry point to target method

Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: **org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderUtils.getEDDSAPublicKeyType()**. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Trace: 
	at parsing org.apache.sshd.common.util.security.SecurityUtils.getEDDSAPublicKeyType(SecurityUtils.java:567)
Call path from entry point to org.apache.sshd.common.util.security.SecurityUtils.getEDDSAPublicKeyType(): 
	no path found from entry point to target method

Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: **org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderUtils.getEDDSASignature()**. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Trace: 
	at parsing org.apache.sshd.common.util.security.SecurityUtils.getEDDSASigner(SecurityUtils.java:556)
Call path from entry point to org.apache.sshd.common.util.security.SecurityUtils.getEDDSASigner(): 
	at org.apache.sshd.common.util.security.SecurityUtils.getEDDSASigner(SecurityUtils.java:555)
	at org.apache.sshd.common.signature.BuiltinSignatures$6.create(BuiltinSignatures.java:98)
	at org.apache.sshd.common.signature.BuiltinSignatures$6.create(BuiltinSignatures.java:95)
	at org.apache.sshd.common.Factory.get(Factory.java:34)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:188)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:141)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:342)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: **org.mozilla.universalchardet.UniversalDetector**. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
Trace: 
	at parsing org.jline.builtins.Nano$Buffer.read(Nano.java:267)
Call path from entry point to org.jline.builtins.Nano$Buffer.read(InputStream): 
	at org.jline.builtins.Nano$Buffer.read(Nano.java:258)
	at org.jline.builtins.Nano$Buffer.open(Nano.java:238)
	at org.jline.builtins.Nano.run(Nano.java:2175)
	at org.jline.reader.impl.LineReaderImpl.editAndAddInBuffer(LineReaderImpl.java:1077)
	at org.jline.reader.impl.LineReaderImpl.editAndExecute(LineReaderImpl.java:3620)
	at org.jline.reader.impl.LineReaderImpl$$Lambda$4858/852570936.apply(Unknown Source)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:665)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:454)
	at org.jline.demo.Graal.main(Graal.java:284)
	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:151)
	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:186)
	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

**STEP III**
```
                <configuration>
                    <skip>false</skip>
                    <imageName>graal</imageName>
                    <mainClass>org.jline.demo.Graal</mainClass>
                    <buildArgs>
                        --no-fallback
                        --report-unsupported-elements-at-runtime
                        --allow-incomplete-classpath
                        -H:ReflectionConfigurationFiles=reflection-config.json
                    </buildArgs>
                </configuration>
```
Compiles without errors and everything seems to work ok except `ttop` command

**Comment by mattirn** [comment]
I have fixed the `java.util.logging` problem and the first `ttop` command problem. Now `ttop` fails like:

graal> ttop
com.oracle.svm.core.jdk.UnsupportedFeatureError: ThreadMXBean methods
	at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:101)
	at com.oracle.svm.core.jdk.SubstrateThreadMXBean.dumpAllThreads(ManagementSupport.java:542)
	at org.jline.builtins.TTop.infos(TTop.java:273)
	at org.jline.builtins.TTop.display(TTop.java:411)
	at org.jline.builtins.TTop.run(TTop.java:197)
	at org.jline.builtins.TTop.ttop(TTop.java:96)
	at org.jline.builtins.Builtins.ttop(Builtins.java:156)
	at org.jline.console.AbstractCommandRegistry.execute(AbstractCommandRegistry.java:66)
	at org.jline.builtins.SystemRegistryImpl.execute(SystemRegistryImpl.java:1086)
	at org.jline.builtins.SystemRegistryImpl.execute(SystemRegistryImpl.java:1132)
	at org.jline.demo.Graal.main(Graal.java:294)

I find it very difficult to solve the problems reported above using available information.
Maybe https://github.com/mattirn/jline3/commit/3906925164d08416daad564150b15937db4e46e1 can be merged to the master as it is now. What do you think?

**Comment by gnodet** [comment]
JMX is not supported in native mode, so the exception is expected I think.  However, it would be nice to catch it and just exit the `ttop` command, or maybe better, to use a substitution and just print an error in native mode.

**Comment by mattirn** [comment]
Ok, I have added catch for `UnsupportedFeatureError` and removed `ttop` command from graal demo.

Actually now we have only two unresolved errors since graal demo does not use  org.apache.sshd.common.util.* .

**Comment by mattirn** [comment]
fixed by pull request: GraalVM #531
